### PR TITLE
The transformation between the inertial coordinates and the Cauchy coordinates in CCE

### PR DESCRIPTION
**Pull request by Sizheng-Ma**
## Proposed changes

Evolving the inertial coordinates as functions of the Cauchy coordinates. The result enables us to establish the transformation between two coordinates at each time step.

### Upgrade instructions

The feature is controlled by `uses_inverse_coordinates` of CCE metavariables. 

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### CODE REVIEW

File: src/Evolution/Systems/Cce/Tags.hpp
Comment (moxcodes) [inline src/Evolution/Systems/Cce/Tags.hpp]: The `name` functions can be omitted because of more recent changes to the tags infrastructure -- without the name, everything that needs a string will just get a string equivalent to the type name.

File: src/Evolution/Systems/Cce/Tags.hpp
Comment (moxcodes) [inline src/Evolution/Systems/Cce/Tags.hpp]: needs an additional slash (`///` for doxygen)

File: src/Evolution/Systems/Cce/Tags.hpp
Comment (moxcodes) [inline src/Evolution/Systems/Cce/Tags.hpp]: Another slash is needed here too.

File: src/Evolution/Executables/Cce/CharacteristicExtract.hpp
Comment (moxcodes) [inline src/Evolution/Executables/Cce/CharacteristicExtract.hpp]: Can we conditionally include the `Cauchy...` jacobian and conformal factors depending on the constexpr boolean in the next commit?

File: src/Evolution/Executables/Cce/CharacteristicExtract.hpp
Comment (moxcodes) [inline src/Evolution/Executables/Cce/CharacteristicExtract.hpp]: Did you mean for this commit and the previous one to be squashed together?

File: src/Evolution/Systems/Cce/Initialize/InitializeJ.hpp
Comment (moxcodes) [inline src/Evolution/Systems/Cce/Initialize/InitializeJ.hpp:163]: Currently this version won't get any documentation. You should probably bring the old dox to this one and update the other specialization with a short explanation of how it differs.

File: src/Evolution/Systems/Cce/Initialize/InverseCubic.cpp
Comment (moxcodes) [inline src/Evolution/Systems/Cce/Initialize/InverseCubic.cpp:47]: For the parts of this function that are identical to the other implementation, does it work to just call `InverseCubic<false>{}(/*shared arguments*/)`? Then this one would only need to include the extra parts associated with the new `.._cauchy_...` coordinates.

File: tests/Unit/Evolution/Systems/Cce/Actions/Test_InitializeFirstHypersurface.cpp
Comment (moxcodes) [inline tests/Unit/Evolution/Systems/Cce/Actions/Test_InitializeFirstHypersurface.cpp]: this can be `if constexpr` -- I think the `constexpr` version is preferable when valid

File: src/Evolution/Executables/Cce/CharacteristicExtract.hpp
Comment (moxcodes) [inline src/Evolution/Executables/Cce/CharacteristicExtract.hpp]: the newline here was added in one commit then removed in this one -- you should choose whichever one clang-format prefers and make the change in only one commit.

Comment (Sizheng-Ma) [reply src/Evolution/Executables/Cce/CharacteristicExtract.hpp]: How do I remove this change from a history commit? @moxcodes

Comment (moxcodes) [reply src/Evolution/Executables/Cce/CharacteristicExtract.hpp]: The way I would do it is to open an interactive rebase with `git rebase -i [parent commit]` and mark the commit to alter with `edit` (instead of `pick`), and then you can `git checkout -p [previous commit hash] [filename]`, it will ask you for each hunk of the commit whether you want to checkout the previous commit's changes. You can just answer 'n' to most of them, and 'y' to the one for this hunk. Once you finish with that, you can `git commit --amend` to put the alterations in the edited commit, then `git rebase --continue` to get back to your branch head.

**Comment by moxcodes** [comment]
fixups look good to me! You can go ahead and squash them in to the original commits (using the `fixup` commend in the git interactive rebase interface).

**Comment by moxcodes** [comment]
@markscheel would you be willing to review this?

**Comment by kidder** [comment]
@Sizheng-Ma this needs a rebase

**Comment by kidder** [comment]
@moxcodes @Sizheng-Ma @markscheel what needs to be done to move this PR forward?

File: src/Evolution/Systems/Cce/Tags.hpp
Comment (markscheel) [inline src/Evolution/Systems/Cce/Tags.hpp]: Naming:  partially-flat Bondi-like coordinates are called "Inertial" here, and are called "CauchyGauge" in `CauchyGaugeC`, `CauchyGaugeD`.  This is confusing What about calling this `PartiallyFlatCartesianCoordinates`, and similar for all the other Tags that refer to quantities in partially-flat Bondi-like coordinates.

Also, it would be nice to have a reference to a equation in the paper that defines this quantity.  And all the other quantities.   Because there are many coordinate systems and they can easily get confusing.

File: src/Evolution/Executables/Cce/CharacteristicExtract.hpp
Comment (markscheel) [inline src/Evolution/Executables/Cce/CharacteristicExtract.hpp]: Maybe `uses_inertial_cartesian_coordinates` ?  Or `uses_partially_flat_cartesian_coordinates` ?   It would be nice if the name here is consistent with the names of the coordinates.

File: src/Evolution/Systems/Cce/Actions/UpdateGauge.hpp
Comment (markscheel) [inline src/Evolution/Systems/Cce/Actions/UpdateGauge.hpp]: Is this template parameter supposed to be the same as what is currently called `EvolutionMetavars::uses_inverse_coordinates` ?  If so, then the two template parameters should have similar names.

**Review by markscheel — COMMENTED** [review]
My main comment is that it would be nice if all the names of the tags, and template parameters agreed with each other, and agreed with what the corresponding coordinates are called in the paper.

File: src/Evolution/Systems/Cce/Tags.hpp
Comment (moxcodes) [inline src/Evolution/Systems/Cce/Tags.hpp]: [nitpick[ you can probably remove this funny newline, probably added by an auto format at some point.

**Review by moxcodes — COMMENTED** [review]
You have my go-ahead to squash, along with the newline I mention here.
Wait to squash until @markscheel has a chance to approve the fixups as well.

File: src/Evolution/Systems/Cce/Initialize/RegisterInitializeJWithCharm.hpp
Comment (markscheel) [inline src/Evolution/Systems/Cce/Initialize/RegisterInitializeJWithCharm.hpp]: Should this template parameter be renamed `uses_partially_flat_cartesian_coordinates`?

File: src/Evolution/Systems/Cce/Initialize/InitializeJ.hpp
Comment (markscheel) [inline src/Evolution/Systems/Cce/Initialize/InitializeJ.hpp]: [minor]  I know this is just a forward declaration, but maybe rename the template param here?

File: src/Evolution/Systems/Cce/Actions/InitializeFirstHypersurface.hpp
Comment (markscheel) [inline src/Evolution/Systems/Cce/Actions/InitializeFirstHypersurface.hpp]: Rename template parameter?

File: src/Evolution/Systems/Cce/OptionTags.hpp
Comment (markscheel) [inline src/Evolution/Systems/Cce/OptionTags.hpp]: Rename template parameter?

File: src/Evolution/Systems/Cce/Actions/InitializeCharacteristicEvolutionVariables.hpp
Comment (markscheel) [inline src/Evolution/Systems/Cce/Actions/InitializeCharacteristicEvolutionVariables.hpp]: Strange indentation here.

**Review by markscheel — COMMENTED** [review]
Looks good. Just a few minor things.

**Comment by markscheel** [comment]
Please squash.

**Review by markscheel — COMMENTED** [review]
Yeah, I think what you did is what Doxygen wants.

At the beginning of the specializations, you say "\brief Abstract base class for an initial hypersurface data generator for Cce."  Could you change this to something like "\brief Abstract base class for an initial hypersurface data generator for Cce, using evolution of inertial coordinates" for the true case, and something similar for the false case?  Then it would be clearer that the two specializations are different.

**Comment by Sizheng-Ma** [comment]
I've edited the specializations.

**Review by moxcodes — COMMENTED** [review]
The additional fixup looks good -- it's an edge case of doxygen that it usually doesn't link properly when trying to link to documentation for template specializations -- I think your fix seems fine as long as the `get_clone`s of the specializations still pass the tests, and it looks like they do.

**Review by moxcodes — APPROVED** [review]

**Review by markscheel — APPROVED** [review]

**Comment by moxcodes** [comment]
@sxs-collaboration/spectre-core-devs This is ready for a last sign-off.

**Comment by kidder** [comment]
clang-tidy error is the one from boost that we haven't figured out how to ignore

**Pull request merged**
